### PR TITLE
Interfaces to the SAS solvers

### PR DESCRIPTION
**Pull request by phchri**
## Fixes #
Provides an interface to the [SAS LP/MILP optimization solvers](https://www.sas.com/en_us/software/optimization.html).

## Summary/Motivation:
Provides an interface to the SAS LP/MILP optimization solvers on both the legacy 9.4 platform (via the saspy package) and on the SAS Viya cloud platform (CAS, via the swat package).

I'm repurposing the solver_io option to choose between the two interfaces, if there is a better way to do this, please advise. The idea is that the interface chooses automatically depending on which package is available.

## Changes proposed in this PR:
- Add the solver interface classes
- Add 52 SAS specific unit tests for the two interfaces, two are currently skipped because the status between LP and MILP is inconsistent (still investigating what's going on there, probably just a status reporting inconsistency on the solver side).

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/solvers/tests/checks/test_SAS.py
Comment (mrmundt) [inline pyomo/solvers/tests/checks/test_SAS.py:76]: Suggestion: move the `skipIf` statement above the class.

File: pyomo/solvers/tests/checks/test_SAS.py
Comment (mrmundt) [inline pyomo/solvers/tests/checks/test_SAS.py:278]: Suggestion: move the `skipIf` statement above the class.

File: pyomo/solvers/tests/checks/test_SAS.py
Comment (mrmundt) [inline pyomo/solvers/tests/checks/test_SAS.py:349]: If this test shows that there is an unexpected failure happening, can the failure be fixed?

File: pyomo/solvers/plugins/solvers/SAS.py
Comment (mrmundt) [inline pyomo/solvers/plugins/solvers/SAS.py:411]: 411 and 412 make more sense after 413's statement concludes; otherwise, they may be getting the full printout twice (once for tee, once for error).

File: pyomo/solvers/plugins/solvers/SAS.py
Comment (mrmundt) [inline pyomo/solvers/plugins/solvers/SAS.py:452]: It doesn't actually look like any `pandas` functions are used; is this correct?

File: pyomo/solvers/plugins/solvers/SAS.py
Comment (mrmundt) [inline pyomo/solvers/plugins/solvers/SAS.py:564]: This is a "magic number" - why this number and not something else? Is it arbitrary? Documented somewhere?

File: pyomo/solvers/plugins/solvers/SAS.py
Comment (mrmundt) [inline pyomo/solvers/plugins/solvers/SAS.py:559]: Is it possible to break this up into smaller, helper functions? It's quite large as-is, and I find it difficult to track all of the logic.

**Review by mrmundt — CHANGES_REQUESTED** [review]
Thanks for this PR! Overall, it looks good; however, we do not (as far as I am aware) have SAS in our testing environment. I will approve GitHub Actions to run for this PR to confirm.

Please look at the two `test_*.yml` files in https://github.com/Pyomo/pyomo/tree/main/.github/workflows

**Comment by mrmundt** [comment]
Confirmed my previous comment: https://github.com/Pyomo/pyomo/actions/runs/5374476640/jobs/9760403183?pr=2886#step:24:8439 <- SAS tests are skipped.

**Comment by phchri** [comment]
> Confirmed my previous comment: https://github.com/Pyomo/pyomo/actions/runs/5374476640/jobs/9760403183?pr=2886#step:24:8439 <- SAS tests are skipped.

Too bad, I was hoping you might have an academic version of at least SAS 9.4 available somewhere.

I ran the unit tests with various SAS versions in my environment, is this good enough for you?

**Comment by mrmundt** [comment]
> 
> Too bad, I was hoping you might have an academic version of at least SAS 9.4 available somewhere.
> 
> I ran the unit tests with various SAS versions in my environment, is this good enough for you?

Unfortunately, no. Because this PR proposes adding the solver into the core offering of Pyomo, we need to ensure its capabilities on our various Python and OS versions before it can be merged.

If you have information on how we might go about obtaining an academic license, we are willing to pursue that.

**Comment by phchri** [comment]
I will investigate how to most easily give you a license or access to machines that run SAS. I'll get back to you as soon as I hear something.

Which institution would a license be for?

**Comment by mrmundt** [comment]
> I will investigate how to most easily give you a license or access to machines that run SAS. I'll get back to you as soon as I hear something.
> 
> Which institution would a license be for?

Thank you! The primary developers are at Sandia National Laboratories, so likely there. Please feel free to email wg-pyomo@sandia.gov so we can discuss further.

File: pyomo/solvers/plugins/solvers/SAS.py
Comment (mrmundt) [inline pyomo/solvers/plugins/solvers/SAS.py:378]: Request here: I'd like to build in the capability for users to tell us where their `sascfg_personal.py` file is located. Can you please add a configuration option called `config_file` or something like that that, if not `None`, gets passed to the creation of the session?

Comment (mrmundt) [reply pyomo/solvers/plugins/solvers/SAS.py:378]: Okay this is now actually important because, surprise, we can run `saspy` on our machines now!

**Comment by mrmundt** [comment]
Good news, @phchri ! We have SAS approval for our test machines.

**Comment by phchri** [comment]
This is great @mrmundt ! I will look into fixing resolving your comments by the end of the week so that we can move ahead with this.

File: pyomo/solvers/plugins/solvers/SAS.py
Comment (phchri) [reply pyomo/solvers/plugins/solvers/SAS.py:378]: One question about this: Do you want the cfgfile to be included in the options dictionary of the solve function or is there a way to include something like this in the SolverFactory constructor?

**Comment by phchri** [comment]
I did the requested changes. It's now possible to add options for saspy.SASsession or swat.CAS when creating the solver factory. All options are just passed through. For example:

SAS94: opt = pyo.SolverFactory('sas', solver_io="_sas94", cfgfile="sasconf.py")
SAS Viya: opt = pyo.SolverFactory('sas', solver_io="_sascas", hostname="HOSTNAME", port="4711", passsword="YOURPASSWORD")

I also fixed a few issues with non-optimal return codes and the version checking.

**Comment by phchri** [comment]
For the unit tests, the test classes have session_options that are passed to the SolverFactory. For SASCAS (SAS Viya) the session options are currently set up to come from environment variables, but this could be adjusted. For SAS 9.4, the default is to use sascfg_personal.py, so currently no session_options are set, but that could be added to point to file or to specify the session options explicitly.

The documentation for the two types of sessions are here:
https://sassoftware.github.io/saspy/api.html
and here:
https://developer.sas.com/apis/swat/python/v1.4.0/generated/swat.cas.connection.CAS.html

File: pyomo/solvers/plugins/solvers/SAS.py
Comment (mrmundt) [inline pyomo/solvers/plugins/solvers/SAS.py:1]: Please add the basic copyright statement to the top of all new files.

File: pyomo/solvers/plugins/solvers/SAS.py
Comment (mrmundt) [inline pyomo/solvers/plugins/solvers/SAS.py:302]: This is a really big function. I recommend breaking it up into some smaller functions.

File: pyomo/solvers/plugins/solvers/SAS.py
Comment (mrmundt) [inline pyomo/solvers/plugins/solvers/SAS.py:566]: Look at `TeeStream`.

**Review by mrmundt — CHANGES_REQUESTED** [review]

**Review by mrmundt — APPROVED** [review]

File: pyomo/solvers/plugins/solvers/SAS.py
Comment (blnicho) [inline pyomo/solvers/plugins/solvers/SAS.py:400]: Typo: mock

File: pyomo/solvers/plugins/solvers/SAS.py
Comment (blnicho) [inline pyomo/solvers/plugins/solvers/SAS.py:405]: Typo: "not implemented"

File: pyomo/solvers/plugins/solvers/SAS.py
Comment (blnicho) [inline pyomo/solvers/plugins/solvers/SAS.py:532]: Typo: pandas

**Review by blnicho — APPROVED** [review]
I found a few small typos in comments and one in an error message but otherwise this looks good.

**Pull request merged**
